pkg/resultx/resultx.go

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package resultx
+
+import (
+	"github.com/arana-db/arana/pkg/proto"
+)
+
+var (
+	_ proto.Result = (*noopResult)(nil)
+	_ proto.Result = (*halfResult)(nil)
+	_ proto.Result = (*fullResult)(nil)
+	_ proto.Result = (*dsOnlyResult)(nil)
+)
+
+type option struct {
+	ds           proto.Dataset
+	id, affected uint64
+}
+
+type Option func(*option)
+
+func WithLastInsertID(id uint64) Option {
+	return func(o *option) {
+		o.id = id
+	}
+}
+
+func WithRowsAffected(n uint64) Option {
+	return func(o *option) {
+		o.affected = n
+	}
+}
+
+func WithDataset(d proto.Dataset) Option {
+	return func(o *option) {
+		o.ds = d
+	}
+}
+
+func New(options ...Option) proto.Result {
+	var o option
+	for _, it := range options {
+		it(&o)
+	}
+
+	if o.ds == nil {
+		if o.id == 0 && o.affected == 0 {
+			return noopResult{}
+		}
+		return halfResult{o.id, o.affected}
+	}
+
+	if o.id == 0 && o.affected == 0 {
+		return dsOnlyResult{ds: o.ds}
+	}
+
+	return fullResult{
+		ds:       o.ds,
+		id:       o.id,
+		affected: o.affected,
+	}
+}
+
+type noopResult struct {
```

Comment:
What's the meaning of `noopResult`? `no operation`?

Reply:
Emm... Yes, means `do noting`.